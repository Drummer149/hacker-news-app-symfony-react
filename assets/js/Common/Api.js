import Firebase from './Firebase';

/**
 * Base API Service to connect to data 
 */
const Api = class ApiService{

  constructor(store = Firebase, version = "v0"){
    this.API_VERSION = version;
    this.DB = Firebase.database();
  };

  // Method to return a promise of an array of story ids from a given endpoint
  getStoryIds(endPoint){
    return new Promise(resolve => {
      this.DB.ref(`${this.API_VERSION}/${endPoint}`).once('value', (storyIds) => {
        resolve(storyIds.val());
      });
    })
  };

  // Method to return a promise of an indivdual item object
  getItem(id){
    return new Promise(resolve => {
      this.DB.ref(`${this.API_VERSION}/item/${id}`).once('value', (item) => {
        resolve(item.val());
      });
    })
  };

};

// Initialise a singleton Api Service 
export default new Api;
