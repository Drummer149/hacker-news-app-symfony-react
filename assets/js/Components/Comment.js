import React from 'react';
import ReactDOM from 'react-dom';
import Api from '../Common/Api';
import InfiniteScroll from 'react-infinite-scroll-component';
import NewsItem from './NewsItem';
import {getHostname, timeSince} from '../Common/Utils';

// Component to display a comment and any sub comments a comment may have
class Comment extends React.Component {

    constructor() {
        super();

        this.state = {
            comment: [],
            hide: false,
            loaded: false,
        };

        this.hideComment = this.hideComment.bind(this);
    }

    componentDidMount() {
        // Get comment from the id provided from the backend
        Api.getItem(this.props.id)
            .then((comment) => {
                this.setState({comment, loaded: true});
            })
    }

    // Method to hide the comment and all sub comments
    hideComment(){
        this.setState({hide : !this.state.hide});
    }

    render() {
        let content = '';
        let display = this.state.hide ? 'none' : 'block';
        let button = this.state.hide ? '+' : '-';

        // Once loaded populate the sub comments, if none return nothing
        if(this.state.loaded && typeof this.state.comment.kids !== 'undefined')
            content = this.state.comment.kids.map((kid) => <Comment key={kid} id={kid} />);
        else if(this.state.loaded && typeof this.state.comment.kids === 'undefined')
            content = '';

        return (<div className="comment">
                    <span>
                        <b>{this.state.comment.by}</b> 
                        {timeSince(this.state.comment.time)} ago 
                        <div className="commentToggle" onClick={this.hideComment}>[{button}]</div>
                    </span>
                    <div className="commentText" style={{display}} >
                        <div dangerouslySetInnerHTML={{__html: this.state.comment.text}} />
                        <div>
                            {content}
                        </div>
                    </div>
                </div>);
    }
}

export default Comment;