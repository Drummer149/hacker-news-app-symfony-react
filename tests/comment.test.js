import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

import React from 'react';
import { shallow } from 'enzyme';
import Api from '../assets/js/Common/Api';
import Comment from '../assets/js/Components/Comment';
import { mockComment } from './mockobjs';

describe('Comment', () => {

  beforeAll(() => {
    const mockGetItem = () => {
      return new Promise(resolve => {
        resolve(mockComment);
      });
    };

    jest.mock('../assets/js/Common/Api', () => {
        return {
          getItem: mockGetItem,
        }
    });
  });

  const initialProps = {
    id: 1
  }

  it('should render', () => {
    const component = shallow(<Comment {...initialProps} />);
    expect(component.exists()).toBe(true);
  });

  it('should render correctly', () => {
    const component = shallow(<Comment {...initialProps} />);
    expect(component).toMatchSnapshot();
  });

  it('should populate the state with the comment coming from the API', () => {
    jest.useFakeTimers();
    const component = shallow(<Comment {...initialProps} />);

    setTimeout(() => {
      expect(component.state().comment).toBe(mockComment);
    }, 1000);
  });

  it('should have method hide and change the state', () => {
    const component = shallow(<Comment {...initialProps} />);
    component.instance().hideComment();
    expect(component.state().hide).toBe(true);
    component.instance().hideComment();
    expect(component.state().hide).toBe(false);
    component.instance().hideComment();
    expect(component.state().hide).toBe(true);
  });
});