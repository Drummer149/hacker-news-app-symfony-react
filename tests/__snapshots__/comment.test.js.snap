// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment should render correctly 1`] = `
<div
  className="comment"
>
  <span>
    <b />
    NaN seconds
     ago
    <div
      className="commentToggle"
      onClick={[Function]}
    >
      [
      -
      ]
    </div>
  </span>
  <div
    className="commentText"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": undefined,
        }
      }
    />
    <div />
  </div>
</div>
`;
